sockeye/translate.py

```diff
@@ -73,27 +63,62 @@ def main():
                                                                                  args.models,
                                                                                  args.checkpoints,
                                                                                  args.softmax_temperature))
-        total_time = 0
-        i = 0
-        for i, line in enumerate(sys.stdin, 1):
-            trans_input = translator.make_input(i, line)
-            logger.debug(" IN: %s", trans_input)
-
-            tic = time.time()
-            trans_output = translator.translate(trans_input)
-            trans_wall_time = time.time() - tic
-            total_time += trans_wall_time
-
-            logger.debug("OUT: %s", trans_output)
-            logger.debug("OUT: time=%.2f", trans_wall_time)
-
-            output_handler.handle(trans_input, trans_output)
-
-        if i != 0:
-            logger.info("Processed %d lines. Total time: %.4f sec/sent: %.4f sent/sec: %.4f", i, total_time,
-                        total_time / i, i / total_time)
-        else:
-            logger.info("Processed 0 lines.")
+        read_and_translate(translator, output_handler, args.source)
+
+
+def read_and_translate(translator: Translator, output_handler: OutputHandler,
+                       source: Optional[str] = None) -> None:
+    """
+
+    :param output_handler: Handler that will write output to a stream.
+    :param translator: Translator that will translate each line of input.
+    :param source: Path to file which will be translated line-by-line if included.
+    """
+    total_time = 0
+    i = 0
+    if source:
+        # Translate from a file.
+        with open(source) as f:
+            i, total_time = _translate(i, output_handler, f.readlines(), total_time, translator, 0)
```

Comment:
reading all input into memory might be dangeruous if translating a HUGE input file. Why not just open a fd or assign to sys.stdin based on args.source/input, like this (also outside of this function):

if args.input is not None:
  input_fd = sockeye.utils.smart_open(args.input)
else:
  input_fd = sys.stdin

with this you can directly call _translate and iterate over it.